public/app/features/datasources/components/EditDataSourceActions.test.tsx

```diff
@@ -318,4 +373,123 @@ describe('EditDataSourceActions', () => {
+      // Core actions should still be rendered
+      expect(screen.getByText('Explore data')).toBeInTheDocument();
+      expect(screen.getByText('Build a dashboard')).toBeInTheDocument();
+    });
+
+    it('should not render favorite button for built-in datasources', () => {
+      config.featureToggles.favoriteDatasources = true;
+      mockUseFavoriteDatasources.mockReturnValue(mockFavoriteHook);
+
+      // Mock built-in datasource
+      const builtInDataSource = { ...mockDataSourceInstance, meta: { ...mockDataSourceInstance.meta, builtIn: true } };
+      mockGetDataSourceSrv.mockReturnValue({
+        getInstanceSettings: jest.fn().mockReturnValue(builtInDataSource),
+        get: jest.fn(),
+        getList: jest.fn(),
+        reload: jest.fn(),
+        registerRuntimeDataSource: jest.fn(),
+      });
+
+      render(<EditDataSourceActions uid="test-uid" />);
+
+      // Should not find any favorite button for built-in datasources
+      expect(screen.queryByRole('button', { name: /favorite|star/i })).not.toBeInTheDocument();
+    });
+
+    it('should render favorite button when feature toggle is enabled and datasource is not built-in', () => {
+      config.featureToggles.favoriteDatasources = true;
+      mockUseFavoriteDatasources.mockReturnValue(mockFavoriteHook);
+      mockFavoriteHook.isFavoriteDatasource.mockReturnValue(false);
+
+      render(<EditDataSourceActions uid="test-uid" />);
+
+      // Should find star icon for non-favorite datasource
+      const favoriteButton = screen.getByRole('button', { name: /add to favorites/i });
+      expect(favoriteButton).toBeInTheDocument();
+
+      // Should have correct aria-label for non-favorite datasource
+      expect(favoriteButton).toHaveAttribute('aria-label', 'Add to favorites');
+    });
+
+    it('should show favorite icon when datasource is favorited', () => {
+      config.featureToggles.favoriteDatasources = true;
+      mockUseFavoriteDatasources.mockReturnValue(mockFavoriteHook);
+      mockFavoriteHook.isFavoriteDatasource.mockReturnValue(true);
+
+      render(<EditDataSourceActions uid="test-uid" />);
+
+      // Should find favorite button for favorited datasource
+      const favoriteButton = screen.getByRole('button', { name: /remove from favorites/i });
+      expect(favoriteButton).toBeInTheDocument();
+
+      // Should have correct aria-label for favorited datasource
+      expect(favoriteButton).toHaveAttribute('aria-label', 'Remove from favorites');
+    });
+
+    it('should add datasource to favorites when star button is clicked', () => {
+      config.featureToggles.favoriteDatasources = true;
+      mockUseFavoriteDatasources.mockReturnValue(mockFavoriteHook);
+      mockFavoriteHook.isFavoriteDatasource.mockReturnValue(false);
+
+      render(<EditDataSourceActions uid="test-uid" />);
+
+      const favoriteButton = screen.getByRole('button', { name: /add to favorites/i });
+      fireEvent.click(favoriteButton);
```

Comment:
Maybe we should use `userEvent` instead of `fireEvent`?
https://testing-library.com/docs/user-event/intro/#differences-from-fireevent

Reply:
the current test file already uses `fireEvent` so I slightly prefer to not break the style here